File: server/test/unit/citations.test.ts
import { describe, it, expect } from 'vitest';
import { extractCitations, scrubUnverifiedCitations } from '../../src/ask/citations.js';

describe('extractCitations', () => {
  it('extracts unique [[slug]] references from text', () => {
    const text = 'See [[idempotent-migrations]] and [[postgres-locks]] for details.';
    expect(extractCitations(text)).toEqual(['idempotent-migrations', 'postgres-locks']);
  });
  it('dedupes repeated citations', () => {
    const text = '[[a]] then [[b]] then [[a]] again.';
    expect(extractCitations(text)).toEqual(['a', 'b']);
  });
  it('ignores non-wiki-link bracket patterns', () => {
    const text = 'Single [brackets] and [[but-not-this stop early.';
    expect(extractCitations(text)).toEqual([]);
  });
  it('only matches slug-shaped contents', () => {
    expect(extractCitations('[[foo bar]]')).toEqual([]);    // space disallowed
    expect(extractCitations('[[FOO]]')).toEqual([]);        // uppercase disallowed
    expect(extractCitations('[[foo-bar]]')).toEqual(['foo-bar']);
  });
});

describe('scrubUnverifiedCitations', () => {
  it('replaces unverified citations with a marker, leaves verified ones alone', () => {
    const verified = new Set(['real-slug']);
    const text = 'See [[real-slug]] and [[fake-slug]] for details.';
    const scrubbed = scrubUnverifiedCitations(text, verified);
    expect(scrubbed).toContain('[[real-slug]]');
    expect(scrubbed).toContain('[unverified citation: fake-slug]');
    expect(scrubbed).not.toContain('[[fake-slug]]');
  });
});

File: cli/internal/citations/citations_test.go
package citations

import (
	"reflect"
	"testing"
)

func TestExtract_UniqueSlugs(t *testing.T) {
	text := "See [[idempotent-migrations]] and [[postgres-locks]] for details."
	got := Extract(text)
	want := []string{"idempotent-migrations", "postgres-locks"}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("Extract = %v, want %v", got, want)
	}
}

func TestExtract_Dedupe(t *testing.T) {
	got := Extract("[[a]] then [[b]] then [[a]] again")
	want := []string{"a", "b"}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("Extract = %v, want %v", got, want)
	}
}

func TestExtract_IgnoresNonWikiLinkBrackets(t *testing.T) {
	got := Extract("Single [brackets] and [[but-not-this stop early.")
	if len(got) != 0 {
		t.Errorf("Extract returned: %v, want empty", got)
	}
}

func TestExtract_SlugShape(t *testing.T) {
	tests := []struct {
		in   string
		want []string
	}{
		{"[[foo bar]]", nil},       // space disallowed
		{"[[FOO]]", nil},           // uppercase disallowed
		{"[[foo-]]", nil},          // trailing hyphen disallowed
		{"[[foo-bar]]", []string{"foo-bar"}},
		{"[[a]]", []string{"a"}},   // single char allowed
		{"[[task-42]]", []string{"task-42"}}, // digits in middle allowed
	}
	for _, tt := range tests {
		t.Run(tt.in, func(t *testing.T) {
			got := Extract(tt.in)
			if len(got) != len(tt.want) {
				t.Errorf("Extract(%q) = %v, want %v", tt.in, got, tt.want)
				return
			}
			for i := range got {
				if got[i] != tt.want[i] {
					t.Errorf("Extract(%q)[%d] = %q, want %q", tt.in, i, got[i], tt.want[i])
				}
			}
		})
	}
}

func TestVerify_PartitionsMatchAndMismatch(t *testing.T) {
	known := map[string]bool{"real-slug": true}
	matched, unmatched := Verify([]string{"real-slug", "fake-slug"}, known)
	if !reflect.DeepEqual(matched, []string{"real-slug"}) {
		t.Errorf("matched = %v", matched)
	}
	if !reflect.DeepEqual(unmatched, []string{"fake-slug"}) {
		t.Errorf("unmatched = %v", unmatched)
	}
}
